### DISCUSSION TITLE
How to achive the  predecessor/successor relationship in Food Packaging sample?

### DISCUSSION THREAD

**Discussion by kentbill** [TimefoldAI/timefold-solver#283] [General]
Hi, there.
I want to implement my scenario based on the Food Packaging example in QuckStarts, a scenario similar to Project Job Scheduling on OptaPlanner - some jobs belong to a project and a job cannot be started before its predecessor jobs are finished, so, if I add these predecessor/successor job relationships to the Food Packaging example, I don't know if it is feasible, **but I think it will cause a Listener Corruption.**
Below is a figure for this situation,
![image](https://github.com/TimefoldAI/timefold-solver/assets/10605984/a65bf0e7-0c71-4576-a10b-3ae3996ef5e0)
From the perspective of the production lines. In Line 1, Job2/Project2 is the previous job of Job2/Project1. In Line 2, Job1/Project1 is the previous job of Job1/Project3	.
From the perspective of project and  job, Job1/Project1 is the predecessor job of  Job2/Project1. 	
When updating the startTime of Job1/Project1, both Job1/Project3 (the next job) and Job2/Project1 (successor job) need to be updated. But when updating Job2/Project1, a listener corruption should be occur because **both its ShadowVariable "line" - Line 1, and "previousJob" - Job2/Project2 are never changed.**	

How to achieve the predecessor/successor relationship of jobs in a project?

Many thanks.

**Comment by kentbill** [comment]
I got an approach, I don't know if it's feasible, I'll try it out. 
In the Food Packaging example, add a Planning Variable - **delay** to the Planning Entity - Job, with a valueRange of 0 - 200 or a suitable number.
In the class StartDateTimeUpdatingVariableListener, method updateStartDateTime, the property **startProductionDateTime** of a job is the endTime of its previous job, plus its **delay** value, and then plus the **cleaning duration**. 
And add a constraint **to penalize a job whose start production time is earlier than the end time of its predecessor job**.

**Comment by nicksi** [comment]
I would say you should implement only one of perspective as a chain of tasks - the Job perspective with delay between tasks as planning variable. In order to mark production line as affected you can use dummy piggyback shadow variable in production line class you touch on each change of task for that line. "Touch" means do before/afterVariableChange call in your listener but keep variable value the same. This would not cause score corruption and have all necessary constraint checks triggered.

**Reply by kentbill** [reply]
Thank you for your suggestion, nicksi.
But I don't understand why we must add a shadow variable to the production line class. Do you mean it will cause a score corruption, if without touching on each change of tasks in that line?
could you please explain it in more detail? 
Thanks!
